pkg/apis/core/v1/defaults.go

```diff
@@ -178,7 +178,9 @@ func SetDefaults_PodSpec(obj *v1.PodSpec) {
 		period := int64(v1.DefaultTerminationGracePeriodSeconds)
 		obj.TerminationGracePeriodSeconds = &period
 	}
-	if obj.SchedulerName == "" {
+
+	// If nodeName is empty, set schedulerName to default scheduler.
+	if obj.SchedulerName == "" && obj.NodeName == "" {
```

Comment:
why is this necessary? doesn't the default scheduler also handle preemption of pods assigned to nodes? is that limited to pods that claim the default scheduler as their scheduler name?

Reply:
If `.spec.NodeName` is not empty, we did not know whether this Pod is scheduled by default scheduler. The scheduler will monitor all `not terminated` and `not assigned` Pod to calculate available resource on the node.

Reply:
please revert this. changing defaults like this can break a HA version-skewed cluster:

1. client creates a pod with a nodeName set via a new API server. API server would persist to etcd without defaulting the scheduler name.
2. client retrieves the pod via an old API server, which returns it with the scheduler name defaulted.
3. client changes something unrelated and submits an update via the new API server. the new API server would reject the request because the incoming pod tried to mutate something in the pod spec (the scheduler name, defaulted by the old API server)

Reply:
Got that! done :)